Add section on the on the Trampoline pod threat (#1)
anfredette/enhancements: enhancements/bpfd-on-ocp.md

  customers away (and would be awkward, messaging-wise: "BPF is great! You can't
  use it!").

#### A Section on mitigating the Trampoline Pod Threat

The [Trampoline Pod Threat](https://drive.google.com/file/d/1ohMZ4DnKMvE2GIu6CO5v4n760N_xvTxE/view?usp=sharing) 
can occur when privileged pods contain elevated kubernetes cluster RBAC permissions. Once a node and

> and ensuing pod is exploited, the privileged credentials can be used to escape ("bounce") a single node's isolation and compromise other nodes in the cluster.

Reviewer: ```suggestion
ensuing pod is exploited, the privileged credentials can be used to escape ("bounce") a single node's
isolation and compromise other nodes in the cluster.
```

This threat can be mitigated by removing functionality from and limiting the permissions

> of daemonsets, the following describes bpfd's strategy for combating this threat.

Reviewer: ```suggestion
of daemonsets. The following describes bpfd's strategy for combating this threat.
```

> Bpfd's deployment includes a daemonset with two containers, bpfd and the bpfd-agent, which is

Reviewer: ```suggestion
Bpfd's deployment includes a daemonset with two containers, bpfd and bpfd-agent, which is
```

//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs/status,verbs=get;update;patch
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs/finalizers,verbs=update
//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
//+kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
//+kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=get;list;watch;create;update;patch;delete
```

bpfd-daemon:

```go
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms/status,verbs=get;update;patch
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms/finalizers,verbs=update
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs,verbs=get;list;watch
//+kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
//+kubebuilder:rbac:groups=core,resources=secrets,namespace=bpfd,verbs=get
```

For the purposes of mitigating the trampoline pod threat we are most concerned with

> with the daemonset permissions, since with privileged daemonsets attackers are guaranteed

Reviewer: ```suggestion
the daemonset permissions, since with privileged daemonsets attackers are guaranteed
```

access to a trampoline pod regardless of which node is compromised. Therefore we will
not analyze the "bounciness" of the bpfd-operator RBAC here.

In the bpfd daemonset, the bpfd-agent container reads user input from the global 
bpfProgramConfig object and then writes it's per node state to it's own 

> bpfProgram Object. It also has permissions to get, list, and watch nodes for node 

Reviewer: In the near future, "I think" it will also need to get, list, and watch pods for pod selection labels if we allow the user to attach to pods with a given label.

Author: +1 Yep It definitely will... Thankfully based on https://static.sched.com/hosted_files/kccnceu2022/35/Trampoline%20Pods_%20Node%20to%20Admin%20PrivEsc%20Built%20Into%20Popular%20K8s%20Platforms.pptx.pdf That's not too dangerous of a permission. 

I'll still add a note here on what the future looks like/ how we plan to keep things documented.

Author: (Also some of those RBAC aren't even in main yet they're in https://github.com/redhat-et/bpfd/pull/372

selection labels and the ability to get secrets **only from the bpfd namespace**
in order to authenticate with private remote image registries.

The privileges granted to the bpfd-daemon service account around bpfd specific APIs are
secure.  The bpfd-agent is unable to write to a bpfProgramConfig object which 
minimizes it's "bounciness" since that's the only bpfd object which could effect other 
nodes in the cluster.  Same goes for the node privilege, which only allows the
service account token to observe nodes and node events. The permission to get 

> secrets is the one with highest risk for "bounciness" however this is mitigated by 

Reviewer: ```suggestion
secrets is the one with highest risk for "bounciness"; however this is mitigated by 
```

  customers away (and would be awkward, messaging-wise: "BPF is great! You can't
  use it!").

> #### A Section on mitigating the Trampoline Pod Threat

Reviewer 2: ```suggestion
#### Mitigating the Trampoline Pod Threat
```

by both the bpfd-daemon and bpfd-operator service accounts.

bpfd-operator: 

```go
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms,verbs=get;list;watch
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs/status,verbs=get;update;patch
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs/finalizers,verbs=update
//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
//+kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
//+kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=get;list;watch;create;update;patch;delete
```

bpfd-daemon:

```go
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms/status,verbs=get;update;patch
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprograms/finalizers,verbs=update
//+kubebuilder:rbac:groups=bpfd.io,resources=bpfprogramconfigs,verbs=get;list;watch
//+kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch

> //+kubebuilder:rbac:groups=core,resources=secrets,namespace=bpfd,verbs=get

Reviewer 3: ```suggestion
//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get
```

Author: Done

```

For the purposes of mitigating the trampoline pod threat we are most concerned
with the daemonset permissions, since with privileged daemonsets attackers are
guaranteed access to a trampoline pod regardless of which node is compromised.
Therefore we will not analyze the "bounciness" of the bpfd-operator RBAC here.

In the bpfd daemonset, the bpfd-agent container reads user input from the global 
bpfProgramConfig object and then writes per node state to its own 
bpfProgram Object. It also has permissions to get, list, and watch nodes for
node selection labels and the ability to get secrets **only from the bpfd namespace**
in order to authenticate with private remote image registries.

The privileges granted to the bpfd-daemon service account around bpfd specific
APIs are secure.  The bpfd-agent is unable to write to a bpfProgramConfig object
which minimizes it's "bounciness" since that's the only bpfd object which could
effect other nodes in the cluster.  Same goes for the node privilege, which only
allows the service account token to observe nodes and node events. The
permission to get secrets is the one with highest risk for "bounciness"; however

> this is mitigated by ensuring the privilege is contained only to the "bpfd"
> namespace.

Reviewer 3: ```suggestion
this is mitigated by requiring the name of the secret thats being got to be know by bpfd ahead of time.
The secret name is provided by the user as part of the operator configuration and that cannot be changed
by the bpfd service account.
```

Author: Done
